### ISSUE TITLE
netatalk2 make weird assumptions around cracklib paths

### ISSUE THREAD

**Issue by VorpalBlade** [Netatalk/netatalk#413]
```
$ afppasswd -a myuser
Enter NEW AFP password: 
/usr/lib64/cracklib_dict.pwd.gz: No such file or directory
Error: error loading dictionary
```

So that is the default from here I presume:

```bash
			if test "$cracklib" = "yes"; then
				cracklib="/usr/$atalk_libname/cracklib_dict"
			fi
```

But I tried overiding it with `--with-cracklib=/usr/share/cracklib/cracklib-small ` (which exists), but then netatalk2 looks for `/usr/share/cracklib/cracklib-small.pwd.gz` (which doesn't).

At this point I'm confused and not quite sure what it wants. Cracklib's man pages also mention `/var/cache/cracklib/cracklib_dict.[hwm|pwd|pwi]` but those don't exist on my system. Yet the command line tool `cracklib-check` seems to work just fine.

**Comment by rdmark** [comment]
It's been a few years since I messed with cracklib now, but I have a vague memory that on Debian, you needed the `cracklib-runtime` package to get those pwd files that contain the actual database (of weak passwords).

**Comment by rdmark** [comment]
@VorpalBlade Do you have suggestions on a concrete way forward to improve the behavior here?

Personally I put little value in the netatalk cracklib integration in this day and age. It's scope is limited to the little used `randnum` UAM, and you're likely not having a large userbase in a public network where password strength is of utmost importance.

In the long run I'd like to deprecate features like cracklib, quota, dropkludge, etc. Maybe AFS, FCL, Webmin, and a few of obsoleted cnid backends.

**Comment by VorpalBlade** [comment]
Sorry, I haven't had time to look into this further (there is no cracklib-runtime or similar package on Arch, and I couldn't find any docs on the arch wiki about cracklib at all, that is about where I stopped).

I changed the package to not build with cracklib instead. Your reasoning makes sense. I needed weird UAMs becuse I'm messing around with classic MacOS (ibook g3 clamshell 300 MHz). I'm not on a big untrusted LAN.

I think the main value of netatalk for most users will be retro computing these days. An easy way to get large files on/off old computers at a reasonable speed. 100 mbit ethernet is way faster than USB 1.1 after all. And there are even older macs without USB. So the imporant thing is probably to keep that sort of functionality.

**Comment by rdmark** [comment]
If your iBook is running OS9 you *should* be able to authenticate with DHX, which is much more secure than randnum. Of course, openssl3 unfortunately broke the OS9 implementation of DHX, but if you can build with openssl1.1, libressl or wolfssl it would be a better choice.

BTW I was looking at your build file. You may not want to enable SLP and Zeroconf at the same time. When I tried to use both in parallel on Debian I ran into issues. See https://github.com/rdmark/netatalk-2.x/issues/7

Please test it though. Maybe there's a way you can get it to work, and have a "universal" service discovery setup for all Mac OS versions ever. :)

**Comment by VorpalBlade** [comment]
> If your iBook is running OS9 you should be able to authenticate with DHX, which is much more secure than randnum. Of course, openssl3 unfortunately broke the OS9 implementation of DHX, but if you can build with openssl1.1, libressl or wolfssl it would be a better choice.

Good to know, I noticed that it wasn't working on OS 9.0 and assumed that was that.

> BTW I was looking at your build file. You may not want to enable SLP and Zeroconf at the same time. When I tried to use both in parallel on Debian I ran into issues. See https://github.com/rdmark/netatalk-2.x/issues/7

Hm, which Mac OS versions are covered by SLP and Zeroconf respectively? Also since they use different ports that probably should work. I'm personally interested in OS 9 mostly, plus early OS X, but I also want to provide a package that is useful for others.

**Comment by rdmark** [comment]
The html manual answers most of your questions, I think.

SLP vs. Zeroconf is described in the Installation chapter: https://netatalk.sourceforge.io/2.2/htmldocs/installation.html#idm105

In practice, only OSX 10.0 and 10.1 require SLP for service discovery. OS9 can still use AppleTalk for service discovery (although AFP connections are always TCP), while OSX 10.2 and later use Zeroconf under the hood for Bonjour. I read somewhere that SLP was a stopgap measure because Zeroconf was still under development and not ready for prime time when Apple wanted to release 10.0.

The configuration chapter has a matrix for the compatibility of the UAMs (Table 3.1) which talks a bit about AppleShare client requirements https://netatalk.sourceforge.io/2.2/htmldocs/configuration.html#authentication

Agreed that it would be nice to have a user friendly matrix with netatalk to AppleShare Client / Mac OS version / UAM compatibility. Perhaps on the wiki. If you have the motivation don't hesitate to get one started. :)

> When do you need netatalk vs afpd for example

Oh, the `netatalk` daemon is just a convenience wrapper for `afpd`. It was added in netatalk3 and doesn't exist in netatalk2. See the man page: https://netatalk.sourceforge.io/3.1/htmldocs/netatalk.8.html

So the answer is: If you have netatalk3 installed, launch `netatalk` to use netatalk. If you have netatalk2 installed, launch `afpd` to use netatalk.

**Comment by VorpalBlade** [comment]
> SLP vs. Zeroconf is described in the Installation chapter: https://netatalk.sourceforge.io/2.2/htmldocs/installation.html#idm105

Ah, must have missed that: "Mac OS 8.5 - Mac OS X 10.4 use SLP to locate AFP servers.". What about system 7 or Mac OS 8 though?

> The configuration chapter has a matrix for the compatibility of the UAMs (Table 3.1) which talks a bit about AppleShare client requirements https://netatalk.sourceforge.io/2.2/htmldocs/configuration.html#authentication

Yeah that is the problem: "For older Macintoshes running Mac OS < X DHCAST128 support exists since AppleShare client 3.8.x.". How do I know what AppleShare client version a given version of Mac OS 8 or 9 has? I did try looking in the AppleTalk and File sharing control panels but it didn't list that.

> So the answer is: If you have netatalk3 installed, launch netatalk to use netatalk. If you have netatalk2 installed, launch afpd to use netatalk.

Sorry I meant `atalkd.service` which *does* exist on netatalk2.

**Comment by rdmark** [comment]
> Ah, must have missed that: "Mac OS 8.5 - Mac OS X 10.4 use SLP to locate AFP servers.". What about system 7 or Mac OS 8 though?

System 6 through OS9 all support AppleTalk service discovery, via the NBP protocol to be exact.

> Yeah that is the problem: "For older Macintoshes running Mac OS < X DHCAST128 support exists since AppleShare client 3.8.x.". How do I know what AppleShare client version a given version of Mac OS 8 or 9 has? I did try looking in the AppleTalk and File sharing control panels but it didn't list that.

Open the System Folder and check the version of the AppleShare extension, like this:

![image](https://github.com/Netatalk/netatalk/assets/2108467/6ac14ab3-bd47-4a7a-8964-315941830d5e)

If in doubt, it's safe to grab an [AppleShare Client installer](https://macintoshgarden.org/search/node/appleshare%20client) with whatever version you want to use. The installer will tell you if the software is compatible with your Mac OS version or not.

> Sorry I meant atalkd.service which does exist on netatalk2.

This section in the manual talks a bit about when to use atalkd: https://netatalk.sourceforge.io/2.2/htmldocs/configuration.html#idm822

I also like to point to the first section of this doc for an overview of DDP vs. DSI: https://github.com/Netatalk/netatalk/blob/branch-netatalk-2-2/doc/DEVELOPER

In short, if you need AppleTalk networking, run atalkd and afpd. If you're happy with TCP networking, run only afpd. As I mentioned, all versions of Classic Mac OS use AppleTalk for service discovery (at least) so it's handy to have running as long as you have at least one Classic Mac OS client. If your network consists solely of OSX / macOS clients, AppleTalk is not necessary.

**Comment by rdmark** [comment]
I should point out that there's a confusing parallel versioning scheme going on between the AppleShare client and AppleShare server products. AFAIK there's no consistent correlation between the two. So in the screenshot about you see AppleShare *client* v3.7.4 is part of AppleShare IP *server* 5.0.2. It throws me off every time.

**Comment by VorpalBlade** [comment]
So OS 8.5-10.4 can use either AppleTalk or SLP for service discovery? Didn't seem to be the case, I needed to enable AppleTalk to get working service discovery on OS 9.0.4 in the Chooser.

This is side tracking this issue at the moment, but I'm considering writing down all of this what I consider a more clear manner (collected in one place, table format), with descriptions of how to check it. Your comments have been invaluable!

(I never did networking (beyond dialup) on Apple back in the day when it was still current, used floppies to move data around mostly, so I lack the institutional knowledge about this.)

**Comment by rdmark** [comment]
Closing this as WontFix as far as Cracklib handling goes.
